Add layer offsets, dynamic object/tile creation, isometric support, and TexturePacker support

timothymtorres opened the pull request:
Consider this PR an unfinished WIP, since I"m trying to get feedback for the final changes that need to be made.  With that said, I'm going to run through a list of the commits made and explain my reasoning.  

### Layer Offsets

Starting with the first few commits, I added layer offsets to all objects.  It was trivial to add and implement this feature.  

In Tiled a user can manipulate a layer by horizontal or vertical offsets.  Neither Berry or Pontiled surprisingly supported this feature.  While I doubt many users will absolutely need this feature, it is definitely nice to have.  (especially useful for isometric layers)  

The only things I have tested layer offsets with are isometric and orthogonal tiles.  At some point polygons and other objects (text/rectangles) will need to be tested as well to confirm it's working for all objects.

 ### Isometric Support

Commit b17c034 adds support for both regular isometric and isometric staggered map types.  This took a while to get working.  Isometric staggered was the most challenging because I was not aware there were extra variables involved.  Notable Tiled has a `staggeraxis` and `staggerindex` variable that appears in the JSON file for only staggered isometric maps.  I used [Simple Tiled Implementation](https://github.com/karai17/Simple-Tiled-Implementation) for Love2D as a reference when making the staggered map type.  The relevant code implementation can be found at this [location](https://github.com/karai17/Simple-Tiled-Implementation/blob/b8e7f42bd645a8d004dacd0382372fb553711468/sti/init.lua#L366).  

I have tested both map types and they appear on screen but the map position is not centered.  Definitely going to need to troubleshoot and fix this at some point.  

### Method `map:getObjects` Improvements

Commit 169d9e4 lets a user filter an object by either objType or name.  Previously you could **only** filter by both.

Commits 678bf93 and 1877e0e combine `getObject` functionality into the `getObjects` method.  Both methods were nearly identical and IMO there was no need to distinguish between plurality of the object(s).  I wrote a commit message in 1877e0e explaining the finer details how this works and why. 


# Major Revision Warning

So all my commits after 1877e0e required me to make a major refactor to Berry.  I don't expect you to accept all the features past this point.  You need to weigh the benefits and whether or not that is the direction you want to take your library.  The above commits I'd argue are a standard necessity that add basic features.  Everything in my next commits add some features users may want, but stuff they definitely might not need.  You may also not like the way I have refactored things, so feel free to say no or request a revision.

With that said, I do like that you have refactored Berry to be a more elegant and smaller version of ponytiled.  Ponytiled had a lot of public methods that an end user was not realistically going to be using.

Since my game needs to have _dynamic_  objects generated **after** a map loads I had to rip out the tile and object creation code and put them into their own functions.  This was a complex endeavor because a lot of that code was coupled together.  

Another problem I came across was that it was hard to decipher the proper scope for variables being used since they all were declared on the same lines.

```lua
    -- Store the flipped states
    local flip, tilesets = {}, data.tilesets
    local tileLayer, objectLayer, layer, tileset, nextTileset, objects, object, image

    if loop1 then
    	if loop2 then
    		if loop3 then
    			for i=1, #loop4 do
    			   -- blah blah blah

```

When in reality the scopes were something like this:

```lua
    -- Store the flipped states
    local tilesets = data.tilesets
    local nextTileset

    if loop1 then
    	if loop2 then
    		local tileLayer, objectLayer
    		if loop3 then
    			local layer
    			local flip = {}
    			local objects
    			for i=1, #loop4 do
    				local object, image
    				-- blah blah blah
```

Onto the next changes!

### Refactor createTile and createObject into their own functions

99b4974 rips out the code in M.new() and puts them into their own functions.  The reason we are doing this is so we can reuse the code when we want to _dynamically_ generate objects later after the map has been loaded.

### Add `map:addLayer` method to Berry

Commit b21679d adds a method for a user to add a new object layer after the map has been loaded.  Note - the added layer is **only an object layer**.  The only argument is a name, which names the object layer.  My reasoning behind this is the only _dynamical_ content a user might want to add is going to be objects.  Adding a tile to a tile layer or adding a new tile layer _after_ the map has been loaded seem like it would be pointless not to mention painfully difficult.  

Initially I had designed the method to handle both tile and object layers and even a option table for an argument to fill the values, but realistically it would be overkill.  Most users are only going to want to add an object at x/y coordinates.  

a57aa34 makes the method also returns the layer after it's added.  This saves a user from having to write another line of code to get the layer or they can just ignore the returned value.  Lua default library does this with some of the functions and I think it's smart.  

### Change module and map names

e5b85a8 changes the module's name to `Map` instead of `M`.  I also was in the process of trying to setup the metatable for the map instance, but found out later that a display group already has a metatable!  I got around this by adding the Map methods directly to the display group via the function `setupDisplayGroup` in 14c4b8a

### Change variable names to be assigned by pairs and ipairs

I noticed you were ignoring the variable assignment of the pairs and ipairs operators which could have saved a few lines of code.  This is one example but there were a few others I encountered.

```lua
			for i=1, #info.data do
				-- GID stands for global tile ID
				gid = info.data[i]
```

Can be rewritten as:

```lua
			for i, gid in ipairs(info.data) do -- GID stands for global tile ID
```

I fixed this with 10638f1 and a few other misc commits.

### Attach data variables to map and layer

Since I had split the createTile and createObject into their own functions the data variables were not in scope.  To remedy this I attached the variables to where they were needed.

```lua
    map.tilesetsDirectory = (tilesetsDirectory and tilesetsDirectory .. '/') or ''
    map.tilesets = data.tilesets
    map.orientation = data.orientation			
    map.staggeraxis = data.staggeraxis
    map.staggerindex = data.staggerindex
    map.tilewidth = data.tilewidth
    map.tileheight = data.tileheight
```

```lua
		-- Apply properties from info
	    layer.name       = info.name
	    layer.type       = info.type
	    layer.alpha      = info.opacity
	    layer.isVisible  = info.visible
            layer.offset_x   = info.offsetx or 0
	    layer.offset_y   = info.offsety or 0
	    layer.properties = info.properties or {} -- Make sure we have a properties table
```

### Change `tileLayer`, `objectLayer`, and `layer`

There was no need to have a different display group for `tileLayer` and `objectLayer`.  Additionally the `layer` variable name was taken by the json.layer.data, so I just renamed this to `info`.  Now `layer` is the sole display group for both `tileLayer` and `objectLayer`.  This made the code shorter and easier to understand IMO.  I kept getting confused which layer was doing what.    These changes happened in 99164c0  but I forgot to mention it.

timothymtorres commented:
### TO DO LIST

- [ ] Get Feedback on PR
- [ ] Add Texture Packer Support
- [ ] Fix isometric map positioning
- [ ] Test layer offsets with polygon, text, and other object types
- [ ] Consider making createTile and createObject into local functions

ldurniat commented on ldurniat/berry.lua:
```diff
@@ -447,484 +468,553 @@ end
+	    layer.type       = info.type
+	    layer.alpha      = info.opacity
+	    layer.isVisible  = info.visible
+		layer.offset_x   = info.offsetx or 0
```
In my opinion you should just apply offset to layer instead for each  individual object. It is simpler solution.
```
layer.x   =  info.offsetx or 0
```
The same apply for y coordinate.
However, this solution requires some change in  `centerAnchor` function. You need replace 
```
object.x = actualCenterX
object.y = actualCenterY
```
with `object.x, object.y = object.parent:contentToLocal( actualCenterX, actualCenterY )`.

timothymtorres replied:
Oh wow.  Hadn’t considered this before.  This would definitely be a good idea to implement.  I completely forgot that you can move the entire position of a display group including all the objects inside it.

timothymtorres replied:
Changes have been made.  Will upload commit tomorrow.

timothymtorres replied:
Commit 26b9778 addresses this.

ldurniat commented on ldurniat/berry.lua:
```diff
@@ -447,484 +468,553 @@ end
-			end	
+				end
+			else  -- must match one
+				if ( object.name == name ) or ( object.type == objType ) then
```
I would modify your condition like that 
````
if ( name and object.name == name ) or ( objType and object.type == objType ) then
````
since you can get objects you don't want e.g. tiles (they don't have `type` property in general).

timothymtorres replied:
Ahhhh, nice catch.  Fixed with 53ede88

ldurniat commented on ldurniat/berry.lua:
```diff
@@ -15,7 +15,7 @@ local json = require 'json'
 --                                  MODULE                                                     --												
```
General note: 

We need stick to one style (notation) e.g. 

`functionName(arq) vs functionName( arq )`

or

```
     end
end
```
vs
```
    end
(blank line)
end
```

I prefer second.
I think you know what I mean:)

timothymtorres replied:
Definitely agree.  I initially kept the same style but as I kept adding more changes and moving blocks of code around things did indeed get a bit sloppy.

timothymtorres replied:
So before I finish the changes to fix this I need to know how you want things done.

Specifically regarding method/variable naming style.

Right now variables and functions both use the same naming styles.  This is mostly camelCase but sometimes things don’t even have capitilzation.  (ie.  `hasbit` `setbit`)

I would really recommend to use camelCase for functions and methods.  Snake_case can be used for variables.  If you have a different preference let me know.

ldurniat replied:
I agree with you.

timothymtorres replied:
Commit fd9f3fe fixes some of this.  Going to finish on knocking out the naming notation for variables and functions soon.

ldurniat replied:
I will check it later. Thanks.

ldurniat commented on ldurniat/berry.lua:
```diff
@@ -447,484 +468,553 @@ end
+--- Create and add tile to layer
+--  
+------------------------------------------------------------------------------------------------
+function Map:createTile(position, gid, layer)
```
Do you test/use it? I'm afraid it can be too complicted to use. The same apply for `:createObject`.

Add `:clone()` function for duplicate objects/tiles can be very useful:) What you think?

timothymtorres replied:
I ripped out the code for the tile and object creation and put them into their own functions for a few reasons.  My primary concern was the ability to reuse these functions to create tiles and objects both before and after the map was loaded.  (this makes content creation more dynamic)

Another reason was mainly code readability.  Before my changes, the map creation function was nested about 6 loops in depth.  This was buried too deep and is generally not good coding standard or practice.  If a function is getting to be too big it's usually a good idea to break it up into smaller functions, which is what I did.  On top of that, careful consideration needs to be made with the scopes of the variables invovled.  

So before my changes the function for the map creation looked like this:

```lua
function Map:new( filename, tilesetsDirectory )
	-- blah blah
	for layer do
		-- blah blah
		if layer.type == 'objectgroup' then
				if object.gid then
					-- blah
					if tileset then
						-- blah
						if imageSheet then
							if findProperty( layer.properties, 'isAnimated' ) or findProperty( object.properties, 'isAnimated' ) then
								-- blah
							end
						end	

						if points then
							if #points > 8 * 2 then
								-- blah
							end
						end
					end	

				elseif object.polygon or object.polyline then 
					-- blah
					for i = 1, #points do
						-- blah
			    	end

				    if object.polygon then 
						if #points > 8 * 2 then 
							-- blah
						end		
				    end
				end

				if image then
					-- blah
					if flip.xy then
						-- blah
					end

					if  findProperty( object.properties, 'hasBody' ) then 
						-- blah
					end	
					-- blah
				end	
			end

		elseif layer.type == 'tilelayer' then
			-- blah
			if gid > 0 then
				-- blah
				if tileset then
					-- blah
					if imageSheet then
						-- blah
					end	

					if image then
						-- blah
					end	
					-- blah
				end	
			end		
		end	
	end 
end
```   

and now it's been simplified to this:

```lua
function Map:new( filename, tilesetsDirectory )
	-- blah blah

	for layer do
		-- blah blah
		if layer.type == 'objectgroup' then
			-- blah blah
			for _, object in ipairs(objects) do
				map:createObject(object, layer)
			end

		elseif layer.type == 'tilelayer' then
			-- blah blah
			for position, gid in ipairs(info.data) do
				if gid > 0 then 
					map:createTile(position, gid, layer) 
				end
			end
		end
	end 
end
```   

Much easier to read and understand IMO.  And it's only nested 3-4 loops deep.  

I believe the `createTile` function might be better suited towards being set as local function instead of being public.  The only reason I made it so, was because I was following the same process as `createObject` when I was refactoring the code.  I doubt many users will want to generate tiles as a tile layer after the map has been loaded.  It is too complex and diffcult to utilize, especially with the postiion and gid.  

For `createObject` on the other hand, this is a different scenario.  Especially if there is TexturePacker support!  The only things a user should need to add the object to a map would be the x and y position and the sprite name.  This would make it simple to add objects to the game.  No dealing with gid's or row/column positions.  Just load the texturepacker sprites into a imagesheet table somewhere and boom.  Ready to go!  At least that's how I'm envisioning it being implemented.

Adding a `:clone()` function might be useful and shouldn't be too difficult to add.  Will probably only be good for cloning objects though as using it on a tile or tilelayer could get complex.

ldurniat replied:
Ok. Thanks for explanation.

ldurniat commented on ldurniat/berry.lua:
```diff
@@ -447,484 +468,553 @@ end
 
-    -- Create group containg all objects including another groups
-    local map  = display.newGroup()
+	local map = setupDisplayGroup( self )
```
Why don't  you  add methods to `map` object (like in Ponytiled) instead using 'setupDisplayGroup' function?

Your method :
```
local Map = {}

function Map:new() 
    ...
    local map = setupDisplayGroup( self )
   ...
end
...
function Map:sort()
     ...
end 
```

vs 

Ponytiled method : 

```
local Map = {}

function Map.new() 
    ...
    local map = setupDisplayGroup( self )
   ...
   function map:sort()
     ...
   end

end 
```

timothymtorres replied:
So I'd say *most* lua libraries in general setup their functions this way: (including ponytiled)

```lua
Library:new()
end

Library:doThis()
end

Library:doThat()
end

Library:doSomething()
end

obj = Library:new(...)
-- obj now has all of Library's methods
```

The way it was initially setup was quite unorthodox.  (I think it's the first time I've ever seen an object inherit its methods that way to be quite honest)

```lua
Library:new()
	obj = ...

	obj:doThis()
	end

	obj:doThat()
	end

	obj:doSomething()
	end
end
```

There was a valid reason for this.  Mainly because you cannot use `setmetatable` on an object that is a display group.  This is due to a display group already having it's own metatable, so the methods need to be inheritted some other way.  One way to get around this is the way it was written initially.  Another way is to just have a `setup` function that attaches the methods to the object like I did.  This is actually what a lot of OO libaries do (middleclass/30log)

As I said in a previous comment, nesting loops too deep is a bad practice.  This problem is magnified if the function itself is large.  Both the deepness and largness were present with the way it was written.  My changes were to resolve this.

On top of all that, I was having issues refactoring the tile and object creation into their own functions with the old way the `M.new` was written.  The order of events for the methods being attacthed caused some unexpected behavior.  

```lua
function M.new(...)
	local map = display.newGroup()

	for i=1, #data.layers do
		...
		if layer.type == 'objectgroup' then
			...
			for object in pairs(objects) do
				map:createObject(...)  -- this is where it would crash

	map:createObject()  -- this function would not load to be used properly

	map:sort()

	map:extend()

	...
```

Because the function `createObject` was created AFTER the objects were getting initialized this resulted in the library crashing.  This is because the `createObject` function does not exist, it has not been loaded by lua yet.  Compare this to:


```lua
function Map:new(...)
	local map = setupDisplayGroup()

	for i=1, #data.layers do
		...
		if layer.type == 'objectgroup' then
			...
			for object in pairs(objects) do
				map:createObject(...)  -- no crashing  :)
	...

Function Map:createObject() -- function loads when library/module is required 
	...
end

Function Map:sort()
	...
end

Function Map:extend()
	...
end
```

The `createObject` function is loaded already and ready to be used.  Hopefully that makes sense.  It took me a while to understand why it was all happening and how to fix it.  Also the scopes of variables were put into their proper function locations where they were needed.

ldurniat replied:
I see now how it should work. I have not built before so robust library like Berry. Again thanks for explanation.

timothymtorres commented on ldurniat/berry.lua:
```diff
@@ -447,484 +468,553 @@ end
+			if self.orientation == 'isometric' then
+
+				image.x = (-1*image.row*self.tilewidth/2) + (image.column*self.tilewidth/2)
+				image.y = (image.column*self.tileheight/2) - (-1*image.row*self.tileheight/2)
```
I'll need to test this out and check, but I think you are right.

timothymtorres replied:
Tested it successfully.  Will upload the commit tomorrow.

timothymtorres replied:
Added via commit df1d539

timothymtorres commented:
I updated my TODO list in the above comment.  Also please know that any commits made to this branch will showup at the bottom of this PR.

timothymtorres commented on ldurniat/berry.lua:
```diff
@@ -447,484 +468,553 @@ end
+--  
+------------------------------------------------------------------------------------------------
+function Map:createTile(position, gid, layer)
+	local tileset
```
Changes have been made.  Will upload commit tomorrow.

timothymtorres commented:
Also I seriously suck at resolving merge conflicts.  I haven't done it too many times in the past...

timothymtorres commented on ldurniat/berry.lua:
```diff
@@ -191,7 +195,7 @@ end
 -- Original code from https://github.com/ponywolf/ponytiled 
------------------------------------------------------------------------------------------------   
+--------------------------------------------------------------------------------   
 local function loadTileset( tileset, dir, tileId )
```
So turns out the `loadTileset` function was doing multiple things depending on the arguments a user would provide.  It would either:

- create a image sheet and return it
- return the directory, width, and height on an image based on tile_id

Functions and methods should do one thing.  (a lot of people call it the [KISS pinciple](https://en.wikipedia.org/wiki/KISS_principle)) The main reason behind this is because the code is a lot easier to reuse when it's in a small state and not coupled or tangled together.  Since I am trying to add TexturePacker functionality to this bit of code, I've had to untangle it and clean it up.  To do that I have taken the function and split it into smaller functions that have a simple purpose.  Now we have,

`getImageSheet(tileset)`
`getTileImage(tileset, tile_id)`

On top of that you can run your logic loop to determine which one you need to use,

```lua
if tileset_has_embedded_images then
  getTileImage(...)
elseif tileset_has_grouped_image then
  getImageSheet(...)
elseif tileset_has_texturepacker_images then
  getTP(...)
```

timothymtorres commented:
Alright I’m finishing up my last changes (95% done) and then I need to do some testing to make sure everything is bug free.  The final changes should be ready by Monday at which point you can test and review everything.

timothymtorres commented:
Alright so here is a rough summary of the changes made.

Changed `buildSequence` to use cache as an argument.  The `sequence_data` for name is now defined via a tile's `name` property.  Before it was set based on the `sequenceName` property.  It is possible to lookup an animation name via an object's gid with the `getAnimationSequence` function.  I set the code to play animations by default when the object is created.  

Texturepacker support is now successfully enabled.  A user has several ways of loading the sprites into Berry.  

1. Place texturepacker sprites and lua files into the tileset_directory for other images
2. Place texturepacker sprites and lua files into their own directory and provide the texturepacker_dir to Map:new
3.  Load each texturepacker files individually.  

For 1 & 2 both the image and lua file names must be the same.  (items.png/items.lua, floor.png/floor.lua)  If the image name is different from the lua file name for whatever reason, the 3rd option is how you load the files instead.  All that is needed is a path for the image file and a path for the lua file.  (so you could have items.png for the image sheet and then objects.lua for the code)

Once the sprites are loaded they can be added to the map very easily.  Just use `map:addSprite` and provide the layer name, image name of the sprite, and the x/y coords for it to go.  No need to mess around with gid's or tile_ids.

One thing I may look into later is applying `map:extend` to the texturepacker objects.

I also gave a map it's own cache variable.  `map.image_cache` that stores pretty much all the gid's and names of objects, tiles, animations, and tilesets.  This should make the lookup *much faster* since it can be done directly.  Also I removed a lot of the code that was deprecated and dead since there is no longer a need to traverse through all the tilesets to find a matching `tile_id`.  I'm very curious to see what the benchmarks are going to show for speed improvements on map creation and loading.  It should be substantial.  

As I said in the discord, I am having issues translating the x/y position of the texture packer sprite for isometric maps.  You may want to take a look at that and see if you can solve it.  Also I did notice that when `isoToScreen` is called in several places, the last argument `offset_y` is never used.  I found that very strange and unusual.  The arguments for the function could be further shortened by just passing in `map` as an argument.  Instead of having to use `isoToScreen(..., map.tile_width, map.tile_height, ...)` every time, you can pass map directly and get the width and height from it in the function.

Last thing for me to nitpick is that I see you have made some really long ternary operations in the code that I can't make sense of.  Generally it's a good idea to encapsulate that into a descriptive local variables that will help both other coders and yourself later on if you come across the same code and forget.  So instead of:

```lua
if objectgroup and objectgroup.objects and #objectgroup.objects > 0 then
  ...
```

can be rewritten as:

```lua
local has_objects_present = objectgroup and objectgroup.objects and #objectgroup.objects > 0
if has_objects_present then
  ...
```

Other than that, I'm glad this is finally being closed to finished.  Took a lot damn longer than I expected.

timothymtorres commented:
Also delete the items.png and items.lua files.  I was using them to test TexturePacker and somehow they must have gotten added.

timothymtorres commented:
[locations.zip](https://github.com/ldurniat/Berry/files/2862704/locations.zip)

These are the files and images I used to profile the code.  Unfortunately, there wasn't much of a performance boost.  I finally realized why.  I had initially assumed that the old version of berry was sorting through _every tile for all tilesets_ in a map.  Turns out this was not the case.  Instead it filters by using `tilset.firstgid > tile.id` and if that is the case, then it searches through the tileset for the gid.  Instead of sorting through thousands of tiles like I initially thought, turns out it was only sorting through a dozen at a time.  So the performance boost is meager.  Meh.  The code refactor still has other benefits so not totally useless.

timothymtorres commented:
[texture_pack.zip](https://github.com/ldurniat/Berry/files/2862732/texture_pack.zip)

This is the texture pack I was testing.  Try placing it somewhere in an isometric map via `map:addSprite`  The x/y position is wrong.  (although it works fine for orthogonal maps)

Relevant code can be found in the `addSprite` function and in the `createObject` function around line `988 - 1015`

The code is

```lua
elseif object.sprite then

	local image_sheet, frame = getImageSheet( map.image_cache, 
											  object.sprite )

	local width, height = getImageSize( map.image_cache, object.sprite )

	image = display.newImageRect( layer, image_sheet, frame, width, height )

	if map.orientation == 'isometric' then

		image.x, image.y = isoToScreen( 
			object.y / map.tile_height, 
			object.x / map.tile_height, 
			map.tile_width, 
			map.tile_height, 
			map.dim.height * map.tile_width * 0.5 
			)
    	image.anchorX, image.anchorY = 0.5, 1   

	elseif map.orientation == 'orthogonal' then 

		image.anchorX, image.anchorY = 0, 1
		image.x, image.y             = object.x, object.y

	end	
```

timothymtorres commented:
Alright it is 100% finished and ready for testing.
